constant interpreter: addresses and memory

marcrasi opened the pull request:
This makes the interpreter able to evaluate functions that do mutation, and this also paves the way for generics support, array support, and string support.

This is a part of the #19579 mega-patch. See there if you're interested in what comes next.

Commit:
constant interpreter: addresses and memory

Adds memory objects and addresses to the constant interpreter, and
teaches the constant interpreter to interpret various instructions that
deal with memory and addresses.

ravikandhadai commented:
@marcrasi  Do you plan to bring in top-level alloc-stack handling via `getSingleWriterAddressValue` function in a different commit?

ravikandhadai commented:
Okay, I see that  `getSingleWriterAddressValue` is in the commit that handles generics.

ravikandhadai commented:
@marcrasi I guess this is the next one that should be merged. Can you rebase this on the new master?

jckarter commented:
It's cool that you were able to handle objects and shared memory locations, and it's awesome that that could let the standard runtime implementations of Array/String/Dictionary "just work". Practically speaking, though, would it be better to represent those common collection types in a more symbolic fashion for compile-time evaluation? Especially if the compile-time evaluator has a cap on number of operations or some other complexity cap, interpreting the standard library implementations at such a low level seems like it could easily blow that cap, or else just be painfully slow to interpret in practice.

marcrasi commented:
The evaluator does have places where we can hook stuff like that in, and in fact it's what we did for some initial support of arrays and strings. You can see it in the huge PR https://github.com/apple/swift/pull/19579/files, if you search `WellKnownFunction` in ConstExpr.cpp. It represents arrays and strings symbolically, and interprets calls to some stdlib functions by operating on those representations.

I haven't created a smaller split patch for that logic yet.

The array/string support is sill very basic. The future work to increase the support will probably add more `WellKnownFunction`s that operate on the symbolic representations.

marcrasi commented:
@ravikandhadai okay, I rebased it and addressed your comments!

ravikandhadai commented:
@marcrasi  Great thanks. Let us wait for @devincoughlin's review.

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci Please test

devincoughlin commented on test/SILOptimizer/pound_assert.swift:
```diff
@@ -154,3 +177,27 @@ func test_CustomStruct() {
+  var myMs = ms
+  addOne(to: &myMs.x.0)
+  addOne(to: &myMs.y)
+  return myMs.x.0 + myMs.x.1 + myMs.y
```
It would be good to make these test a bit more comprehensive. In particular, it would be good to make sure there is not unexpected aliasing (for example, between `myMs` and `ms` above).

It would also be good to test the case where you are storing an aggregate to a projection. Something like:

```
struct Foo {
  var x: Int 
  var y: Int

  init() {
   x = 7
   y = 8
  }
}

struct HasFoo {
  var f = Foo()
}

func foo() -> Bool {
  var h = HasFoo()
  var hp = h
  h.f.x = 22
  hp.f = h.f
  return (hp.f.x == 7)
}

func bar() {
  #assert(foo())
}
```

marcrasi replied:
I added a test for undesired aliasing, and a test for storing aggregate to projection.

devincoughlin commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -523,6 +577,152 @@ SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
+  auto *memoryObject = it->second.getAddressValue(accessPath);
+
+  // If this is a direct store to tracked memory object, just update it.
+  if (accessPath.empty()) {
```
Is this needed? It seems like the base case of updateIndexedElement() should already handle this, regardless of whether it is a derived address.

marcrasi replied:
Yeah, seems removable. I removed it.

devincoughlin commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -523,6 +577,152 @@ SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
+  }
+
+  // If we have an uninit memory, then scalarize it into an aggregate to
+  // continue.  This happens when memory objects are initialized piecewise.
```
Is it possible to write test that fails when this scalarization is removed?

marcrasi replied:
Yep, I added a SIL test for it!

devincoughlin commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -174,6 +191,39 @@ SymbolicValue ConstExprFunctionState::computeConstantValue(SILValue value) {
+  // value of the memory.  In the latter case, the call must be the only
+  // store to the address so that the memory object can be computed by
+  // recursively processing the allocation and call instructions in a
+  // demand-driven fasion.
```
Typo: "fasion"

devincoughlin commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -539,6 +739,21 @@ ConstExprFunctionState::evaluateFlowSensitive(SILInstruction *inst) {
+
+  // If this is a deallocation of a memory object that we may be tracking,
+  // remove the memory from the set.  We don't *have* to do this, but it seems
+  // useful for hygiene.
```
Is there some kind of assertion that could be added to verify this hygiene? For example, at function exit?

marcrasi replied:
Now that my attention has been drawn to this code-comment, I realize that the code-comment and the code is wrong. It's left over from an earlier memory implementation where the memory values actually got stored in `calculatedValues`.

I have changed the code to do nothing here, and updated the comment to explain. We could do something a bit more elaborate, like add a new "freed" state to the memory and make it an assertion failure whenever you try to do any operations to "freed" memory.

devincoughlin commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -553,6 +768,23 @@ ConstExprFunctionState::evaluateFlowSensitive(SILInstruction *inst) {
+  }
+
+  // Copy addr is a load + store combination.
+  if (auto *copy = dyn_cast<CopyAddrInst>(inst)) {
```
It would be good to add a test for this functionality. This should probably be a SIL-level test so that you can guarantee that the test continues to test the functionality even if SILGen changes its strategy about what is emitted as a CopyAddr.

marcrasi replied:
Added a SIL test!

devincoughlin reviewed (commented):
It is great to handling of address projections and mutation! This looks good to me with the comments inline addressed.

ravikandhadai commented:
@marcrasi  I guess this PR is good to go, once Devin's comments are fixed.

ravikandhadai commented on lib/SIL/SILConstants.cpp:
```diff
@@ -120,9 +139,30 @@ SymbolicValue::cloneInto(ASTContext &astContext) const {
+SymbolicValueMemoryObject *
+SymbolicValueMemoryObject::create(Type type, SymbolicValue value,
+                                  ASTContext &astContext) {
+  auto result = astContext.Allocate(sizeof(SymbolicValueMemoryObject),
```
@marcrasi a minor nitpick I just noticed. It is better to use `auto *` here.

ravikandhadai commented on lib/SIL/SILConstants.cpp:
```diff
@@ -120,9 +139,30 @@ SymbolicValue::cloneInto(ASTContext &astContext) const {
+  auto result = astContext.Allocate(sizeof(SymbolicValueMemoryObject),
+                                    alignof(SymbolicValueMemoryObject));
+  new (result) SymbolicValueMemoryObject(type, value);
+  return (SymbolicValueMemoryObject*)result;
```
@marcrasi  space before '*'

marcrasi commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -523,6 +577,152 @@ SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
+///
+static bool updateIndexedElement(SymbolicValue &aggregate,
+                                 ArrayRef<unsigned> indices,
+                                 SymbolicValue scalar, Type type,
```
changed to `newElement`

marcrasi commented:
I have rebased this and addressed all the comments.

While addressing comments I also noticed that I was supposed to do [this TODO](https://github.com/apple/swift/blob/30bda999ddf5b09345dbe1261ee8188cb0b0f1a8/lib/SILOptimizer/Utils/ConstExpr.cpp#L83) in this PR, so I attempted to do it. The resulting assertion failed, because we actually do store addresses in `calculatedValues`. It turns out that [the comment](https://github.com/apple/swift/blob/30bda999ddf5b09345dbe1261ee8188cb0b0f1a8/lib/SILOptimizer/Utils/ConstExpr.cpp#L72) that led us to introduce this assertion was actually misleading. It meant to say that we don't store memory values in `calculatedValues`. So I deleted the assertion and clarified the comment.

Sorry about the slowness. I've been busy. I have everything that I need to start working on the next PR in the series (https://github.com/apple/swift/pull/19662), but I think I'll keep being busy until the end of the year. Is it okay if I delay working on the next parts until early Jan? I think it'll take a few weeks of back and forth to get the rest of constexpr merged once I start working on it again, so we'd be looking at finishing it in early Feb.

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please Test and Merge

marcrasi commented:
Oh no, looks like some of the `pound_assert.sil` tests fail when `Int` is 32-bit. What is a good way to handle that?

devincoughlin commented:
@marcrasi You should be able to fix the 32-bit failures by explicitly making the tuple you create in piecewiseInit() to be Int64 instead of Int.
